src/java/htsjdk/samtools/CRAMContainerStreamWriter.java

```diff
@@ -0,0 +1,460 @@
+    private final String cramID;
+    private final OutputStream outputStream;
+    private ReferenceSource source;
+
+    private final List<SAMRecord> samRecords = new ArrayList<SAMRecord>();
+    private ContainerFactory containerFactory;
+    private int refSeqIndex = REF_SEQ_INDEX_NOT_INITIALIZED;
+
+    private static final Log log = Log.getInstance(CRAMContainerStreamWriter.class);
+
+    private boolean preserveReadNames = true;
+    private QualityScorePreservation preservation = null;
+    private boolean captureAllTags = true;
+    private Set<String> captureTags = new TreeSet<String>();
+    private Set<String> ignoreTags = new TreeSet<String>();
+
+    private CRAMIndexer indexer;
+    private long offset;
+
+    /**
+     * Create a CRAMContainerStreamWriter for writing SAM records into a series of CRAM
+     * containers on output stream, with an optional index.
+     *
+     * @param outputStream where to write the CRAM stream.
+     * @param indexStream where to write the output index. Can be null if no index is required.
+     * @param source reference source
+     * @param samFileHeader {@link SAMFileHeader} to be used. Sort order is determined by the sortOrder property of this arg.
+     * @param cramId used for display in error message display
+     */
+    public CRAMContainerStreamWriter(
+            final OutputStream outputStream,
+            final OutputStream indexStream,
+            final ReferenceSource source,
+            final SAMFileHeader samFileHeader,
+            final String cramId) {
+        this.outputStream = outputStream;
+        this.samFileHeader = samFileHeader;
+        this.cramID = cramId;
+        this.source = source;
+        containerFactory = new ContainerFactory(samFileHeader, recordsPerSlice);
+        if (indexStream != null) {
+            indexer = new CRAMIndexer(indexStream, samFileHeader);
+        }
+    }
+
+    /**
+     * Write an alignment record.
+     * @param alignment must not be null
+     */
+    public void writeAlignment(final SAMRecord alignment) {
+        if (shouldFlushContainer(alignment)) {
+            try {
+                flushContainer();
+            } catch (IOException e) {
+                throw new RuntimeIOException(e);
+            } catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        updateReferenceContext(alignment.getReferenceIndex());
+
+        samRecords.add(alignment);
+    }
+
+    /**
+     * Write a CRAM file header and SAM header to the stream.
+
+     * @param textHeader
+     */
+    public void writeHeader(final String textHeader) {
+        // TODO: header must be written exactly once per writer life cycle.
+        final SAMFileHeader header = new SAMTextHeaderCodec().decode(new StringLineReader(textHeader), (cramID != null ? cramID : null));
+        offset = CramIO.writeHeader(cramVersion, outputStream, header, cramID);
+    }
+
```

Comment:
Good idea. Once I do that I no longer need the text version here, so I'll just replace it and let CRAMFileWriter do the conversion since it has to implement the text based once.